clients/da-vinci-client/src/test/java/com/linkedin/davinci/blobtransfer/TestP2PFileTransferClientHandler.java

```diff
@@ -0,0 +1,433 @@
+    response1.headers().add("Content-Length", "5");
+    response1.headers().add(BLOB_TRANSFER_TYPE, BlobTransferType.FILE);
+    // response 2
+    DefaultHttpResponse response2 = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
+    response2.headers().add("Content-Disposition", "filename=\"test_file2.txt\"");
+    response2.headers().add("Content-Length", "10");
+    response2.headers().add(BLOB_TRANSFER_TYPE, BlobTransferType.FILE);
+    // content
+    HttpContent chunk1 = new DefaultLastHttpContent(Unpooled.copiedBuffer("12345", CharsetUtil.UTF_8));
+    HttpContent chunk2 = new DefaultHttpContent(Unpooled.copiedBuffer("67890", CharsetUtil.UTF_8));
+    HttpContent chunk3 = new DefaultLastHttpContent(Unpooled.copiedBuffer("13579", CharsetUtil.UTF_8));
+
+    // End of all file transfer
+    DefaultHttpResponse endOfTransfer = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
+    endOfTransfer.headers().add(BLOB_TRANSFER_STATUS, BLOB_TRANSFER_COMPLETED);
+
+    ch.writeInbound(response1);
+    ch.writeInbound(chunk1);
+    ch.writeInbound(response2);
+    ch.writeInbound(chunk2);
+    ch.writeInbound(chunk3);
+    ch.writeInbound(endOfTransfer);
+    inputStreamFuture.toCompletableFuture().get(1, TimeUnit.MINUTES);
+
+    // verify the content is written to the disk
+    BlobTransferPayload payload = new BlobTransferPayload(baseDir.toString(), TEST_STORE, TEST_VERSION, TEST_PARTITION);
+    Path dest = Paths.get(payload.getPartitionDir());
+    Assert.assertTrue(Files.exists(dest));
+    Assert.assertTrue(Files.isDirectory(dest));
+    Path file1 = dest.resolve("test_file1.txt");
+    Assert.assertTrue(Files.exists(file1));
+    Assert.assertEquals(Files.size(file1), 5);
+    Path file2 = dest.resolve("test_file2.txt");
+    Assert.assertTrue(Files.exists(file2));
+    Assert.assertEquals(Files.size(file2), 10);
+  }
+
+  @Test
+  public void testSingleMetadataTransfer()
+      throws JsonProcessingException, ExecutionException, InterruptedException, TimeoutException {
+    BlobTransferPartitionMetadata expectedMetadata = new BlobTransferPartitionMetadata();
+    expectedMetadata.setTopicName(TEST_STORE + "_v" + TEST_VERSION);
+    expectedMetadata.setPartitionId(TEST_PARTITION);
+    InternalAvroSpecificSerializer<PartitionState> partitionStateSerializer =
+        AvroProtocolDefinition.PARTITION_STATE.getSerializer();
+    OffsetRecord offsetRecord = new OffsetRecord(partitionStateSerializer);
+    offsetRecord.setOffsetLag(1000L);
+    expectedMetadata.setOffsetRecord(ByteBuffer.wrap(offsetRecord.toBytes()));
+
+    ObjectMapper objectMapper = new ObjectMapper();
+    String metadataJson = objectMapper.writeValueAsString(expectedMetadata);
+    byte[] metadataBytes = metadataJson.getBytes(CharsetUtil.UTF_8);
+
+    DefaultHttpResponse metadataResponse = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
```

Comment:
If we use `DefaultFullHttpResponse`, we don't need `metadataChunk` and `endOfTransfer`, correct?